shared/src/balancer_event_handler.rs

```diff
@@ -54,15 +117,127 @@ impl PoolSpecialization {
             0 => Ok(Self::General),
             1 => Ok(Self::MinimalSwapInfo),
             2 => Ok(Self::TwoToken),
-            t => Err(anyhow!("Invalid PoolSpecialization value {}", t)),
+            t => Err(anyhow!("Invalid PoolSpecialization value {} (> 2)", t)),
         }
     }
 }
 
-#[derive(Default)]
-pub struct BalancerPools {}
-
 impl BalancerPools {
+    fn try_upgrade(&mut self, pool_builder: WeightedPoolBuilder) {
+        match pool_builder.clone().try_into() {
+            Ok(weighted_pool) => {
+                let token_registration = pool_builder.tokens_registration.expect("known to exist");
```

Comment:
nit: why not use pool_id and tokens from the already built weighted pool directly?

Reply:
I was struggling a lot with the borrow checker when refactoring this and this is likely left over from several different approaches.

Reply:
You could clone the weighted pool instead of the builder if needed for the borrow checker (you are already cloning the builder, so it's at the end equivalent). If you have a concrete error, I'm sure we can help out.